src/main.rs

```
@@ -0,0 +1,216 @@
+        #[clap(flatten)]
+        build_run: BuildRunArgs,
+    },
+}
+
+/// Arguments common to all subcommands.
+#[derive(Args, Debug)]
+struct CommonArgs {
+    /// Input file (.desc)
+    input: String,
+
+    /// Optionally write the CUDA code to a file (if not provided, uses default naming)
+    #[arg(short, long)]
+    output: Option<String>,
+}
+
+/// Arguments only applicable to Build and Run.
+#[derive(Args, Debug)]
+struct BuildRunArgs {
+    /// Specify CUDA architecture (e.g., sm_75, sm_80)
+    #[arg(long, default_value = "sm_75")]
+    arch: String,
+
+    /// Optimization level for nvcc (0-3)
+    #[arg(long, default_value = "3")]
+    optimize: u8,
+
+    /// Additional flags to pass directly to nvcc
+    #[arg(long, default_value = "")]
+    nvcc_flags: String,
+}
+
+/// Checks if a command exists using the which crate
+fn command_exists(cmd: &str) -> bool {
+    which(cmd).is_ok()
+}
+
+fn generate_cuda(input: &str) -> Result<String> {
+    compile(input, None)
+        .map_err(|_| anyhow::anyhow!("Descend compilation failed for input '{}'", input))
+}
+
+fn write_cuda_file(cuda_code: &str, filename: &str) -> Result<()> {
+    write(filename, cuda_code).with_context(|| format!("Error writing CUDA file {}", filename))
+}
+
+fn build_cuda(
+    cuda_file: &str,
+    executable: &str,
+    optimize: u8,
+    arch: &str,
+    nvcc_flags: &str,
+) -> Result<()> {
+    let mut nvcc_cmd = Command::new("nvcc");
+    nvcc_cmd
+        .arg(cuda_file)
+        .arg("-o")
+        .arg(executable)
+        .arg(format!("-O{}", optimize))
+        .arg("-I")
+        .arg("cuda-examples/")
+        .args(nvcc_flags.split_whitespace());
+    if arch != "none" {
+        nvcc_cmd.arg(format!("-arch={}", arch));
+    }
+    debug!("Running NVCC command: {:?}", nvcc_cmd);
+    let output = nvcc_cmd
+        .output()
+        .with_context(|| "Failed to run nvcc command")?;
+    if !output.status.success() {
+        return Err(anyhow::anyhow!(
```

I think `anyhow` could be a custom error. When calling `compile` in l.84 we could simply bubble up the error.
How this should work will probably be more clear after PR #31 has been merged.

---

I introduced two custom errors called `NVCCError` and `ExecutableError` in error.rs and removed the use of anyhow. I also added a `std::fmt::Display` implementation for ErrorReported. This should allow us to bubble up errors in a more controlled way.